lib/rules/require-named-capture-group.js

```diff
@@ -0,0 +1,90 @@
+/**
+ * @fileoverview Rule to enforce requiring named capture groups in regular expression.
+ * @author Pig Fang <https://github.com/g-plane>
+ */
+
+"use strict";
+
+//------------------------------------------------------------------------------
+// Requirements
+//------------------------------------------------------------------------------
+
+const regexpp = require("regexpp");
+
+//------------------------------------------------------------------------------
+// Helpers
+//------------------------------------------------------------------------------
+
+const parser = new regexpp.RegExpParser();
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = {
+    meta: {
+        type: "suggestion",
+
+        docs: {
+            description: "enforce using named capture group in regular expression",
+            category: "Best Practices",
+            recommended: false,
+            url: "https://eslint.org/docs/rules/require-named-capture-group"
+        },
+
+        schema: [],
+
+        messages: {
+            required: "Capture group '{{group}}' in regular expression should be named."
+        }
+    },
+
+    create(context) {
+
+        /**
+         * Function to check regular expression.
+         *
+         * @param {string} regex The regular expression to be check.
+         * @param {ASTNode} node AST node which contains regular expression.
+         * @returns {void}
+         */
+        function checkRegex(regex, node) {
+            const ast = parser.parsePattern(regex);
+
+            regexpp.visitRegExpAST(ast, {
+                onCapturingGroupEnter(group) {
+                    if (!group.name) {
+                        context.report({
+                            node,
+                            messageId: "required",
```

Comment:
Giving `loc` is better because multiple reports can be reported at the same location currently. It may be the following code (though I didn't run it):

```suggestion
                            loc: {
                                start: {
                                    column: node.loc.start.column + group.start,
                                    line: node.loc.start.line
                                },
                                end: {
                                    column: node.loc.start.column + group.end,
                                    line: node.loc.start.line
                                }
                            },
                            messageId: "required",
```